fix(css): update opacity page (#31725)
mdn/content: files/en-us/web/css/opacity/index.md

`opacity` applies to the element as a whole, including its contents, even though the value is not inherited by child elements. Thus, the element and its children all have the same opacity relative to the element's background, even if they have different opacities relative to one another.

To change the opacity of a background only, use the {{cssxref("background")}} property with a {{cssxref("color_value", "color value")}} that allows for an alpha channel. For example:

```css
background: rgb(0 0 0 / 40%);
```

> When `opacity` value is set to `0`, the element and all of its children appear invisible, but they are still part of the DOM. That means they still register [pointer events](/en-US/docs/Web/API/Pointer_events) and if the elements are in a tabbing order, they do get focus. For good usability make sure to make such elements visible when they receive user interactions or use the CSS [`pointer-events`](/en-US/docs/Web/CSS/pointer-events) property to disable pointer events and [`tab-index="-1"`](/en-US/docs/Web/HTML/Global_attributes/tabindex) to take element out of tab order.

Reviewer: ```suggestion
When `opacity` value is set to `0`, the element and all of its children appear invisible, but they are still part of the DOM. That means they still register [pointer events](/en-US/docs/Web/API/Pointer_events) and, if the elements are in a tabbing order, they do get focus. For good usability, make sure to make such elements visible when they receive user interactions or use the CSS [`pointer-events`](/en-US/docs/Web/CSS/pointer-events) property to disable pointer events and take the element out of the tab order by disabling with the `disabled` attribute or setting [`tab-index="-1"`](/en-US/docs/Web/HTML/Global_attributes/tabindex) for non-form-related interactive elements.
```
not sure if the above verbiage is what you want to go for in the end, but if i were to disable a fieldset of form controls, i would disable the fieldset, not alter tab indexes, so recommending including that, however you want to include it.

> In case of applying [transitions](/en-US/docs/Web/CSS/CSS_transitions) related to visibility on elements, the `opacity` property is required to be set in [`@starting-style`](/en-US/docs/Web/CSS/@starting-style) at-rule as well:

Reviewer: When transitioning the opacity of elements as you add them to the page when content was formerly hidden with `visibility: hidden`, `display: none`, or `content-visibility: hidden` you want to include both a `@starting-style` and `transition-behavior: allow-discrete`.  

In this example, when hidden is removed, it will be dispayed and the transition will occur as the display none gets removed immmediately. When it gets hidden, no transition will be visible as the display none happens right away. We should either mention this, or explain how to transition in both directions.

With allow-discrete, when animating display from `none` to a visible display value, the value will flip to the visible value at 0% of the transitions and will to none at 100% in the reverse direction so the transition is visible throughout. 

Also, in this example, the opacity isn't necessary on the hidden as it's not displayed, so it's not in the DOM and the related styles aren't applied i don't think.

Author: As this is getting bigger I've moved this to a sub section. And added more information.

There are more nuances to transitioning `display: none` and `@starting-style`. Can we keep this small mentioning only that the `@starting-style` and `transition-behavior` are required for transitioning opacity and pointing readers towards the actual pages?

> Also, in this example, the opacity isn't necessary on the hidden as it's not displayed, so it's not in the DOM and the related styles aren't applied i don't think.

It is needed to transition to `.hidden`: https://jsfiddle.net/971apcfy/

To change the opacity of a background only, use the {{cssxref("background")}} property with a {{cssxref("color_value", "color value")}} that allows for an alpha channel. For example:

```css
background: rgb(0 0 0 / 40%);
```

When `opacity` value is set to `0`, the element and all of its children appear invisible, but they are still part of the DOM. That means they still register [pointer events](/en-US/docs/Web/API/Pointer_events) and, if the elements are in a tabbing order, they do get focus. For good usability, make sure to make such elements visible when they receive user interactions or use the CSS [`pointer-events`](/en-US/docs/Web/CSS/pointer-events) property to disable pointer events and take the element out of the tab order by disabling with the `disabled` attribute or setting [`tab-index="-1"`](/en-US/docs/Web/HTML/Global_attributes/tabindex) for non-form-related interactive elements.

Using `opacity` with a value other than `1` places the element in a new [stacking context](/en-US/docs/Web/CSS/CSS_positioned_layout/Understanding_z-index/Stacking_context).

> Opacity alone should not be used to provide information to screen readers. Use the HTML [`hidden`](/en-US/docs/Web/HTML/Global_attributes/hidden) attribute, CSS [`visibility`](/en-US/docs/Web/CSS/visibility), or CSS [`display`](/en-US/docs/Web/CSS/display) style properties. It's best to avoid using [`aria-hidden`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden) attribute, but if the element is hidden with opacity then hide it from screen readers too.

Reviewer: ```suggestion
Opacity alone should not be used to provide information to screen readers. Use the HTML [`hidden`](/en-US/docs/Web/HTML/Global_attributes/hidden) attribute, CSS [`visibility`](/en-US/docs/Web/CSS/visibility), or CSS [`display`](/en-US/docs/Web/CSS/display) style properties. It's best to avoid using [`aria-hidden`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden) attribute, but if the element is hidden with opacity, then hide it from screen readers as well.
```

### Transitioning opacity

> When [transitioning](/en-US/docs/Web/CSS/CSS_transitions) the opacity of elements as you add them to the page when content was formerly hidden with [`visibility: hidden`](/en-US/docs/Web/CSS/visibility#hidden), [`display: none`](/en-US/docs/Web/CSS/display#none), or [`content-visibility: hidden`](/en-US/docs/Web/CSS/content-visibility#hidden) you need to include both [`@starting-style`](/en-US/docs/Web/CSS/@starting-style) and [`transition-behaviour: allow-descrete`](/en-US/docs/Web/CSS/transition-behavior#allow-discrete):

Reviewer: ```suggestion
When [transitioning](/en-US/docs/Web/CSS/CSS_transitions) the opacity of elements as you add them to the page when content was formerly hidden with [`visibility: hidden`](/en-US/docs/Web/CSS/visibility#hidden), [`display: none`](/en-US/docs/Web/CSS/display#none), or [`content-visibility: hidden`](/en-US/docs/Web/CSS/content-visibility#hidden), you need to include both a [`@starting-style`](/en-US/docs/Web/CSS/@starting-style) and [`transition-behaviour: allow-discrete`](/en-US/docs/Web/CSS/transition-behavior#allow-discrete):
```

```css
.card {
  transition:
    opacity 5s,
    display 5s;
  background-color: orange;

  transition-behavior: allow-discrete;
  @starting-style {
    opacity: 0;
  }
}

.card.hidden {
  display: none;
  opacity: 0;
}
```

> To enable first-style transitions, `@starting-style` rules are needed. In above code, setting `opacity: 0` in `@starting-style` provides starting point for the transition when the element receives initial style update. For more details refer [`@starting-style`](/en-US/docs/Web/CSS/@starting-style) page.

Reviewer: ```suggestion
To enable first-style transitions, `@starting-style` rules are needed. In the above code, setting `opacity: 0` in `@starting-style` provides a starting point for the transition when the element receives its initial style update. For more details, see [`@starting-style`](/en-US/docs/Web/CSS/@starting-style).
```

> Setting `transition-behaviour: allow-descrete` is required to transition to `display: none`. For more details refer [`transition-behaviour`](/en-US/docs/Web/CSS/transition-behavior) page.

Reviewer: ```suggestion
Setting `transition-behavior: allow-discrete` is required to transition to `display: none`. See the [`transition-behavior`](/en-US/docs/Web/CSS/transition-behavior) property for more details.
```
